File: src/nuclearbot/gui/components/commands/CommandEditPanel.java
package nuclearbot.gui.components.commands;

import nuclearbot.gui.NuclearBotGUI;
import nuclearbot.gui.commands.UserCommandManager;
import nuclearbot.gui.utils.VerticalLayout;

import javax.swing.*;
import java.awt.*;
import java.awt.event.FocusAdapter;
import java.awt.event.FocusEvent;
import java.awt.event.ItemEvent;

/*
 * Copyright (C) 2017 NuclearCoder
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * The GUI panel for the command edition panel.<br>
 * <br>
 * NuclearBot (https://github.com/NuclearCoder/nuclear-bot/)<br>
 *
 * @author NuclearCoder (contact on the GitHub repo)
 */
public class CommandEditPanel extends JPanel {

    private static final String HELP_TEXT = "The convention for usage format is the following:\n"
            + "- angle brackets for required arguments;\n"
            + "- square brackets for optional arguments;\n"
            + "but the GUI creator doesn't support optional arguments yet.\n"
            + "Please only write arguments without the command in the usage field.\n\n"
            + "To include arguments in the response, use \"$n\" or \"{$n}\", where n is the argument index.\n"
            + "The zero-th argument is the sender's username.\n"
            + "For instance, to create a hug command, put \"!hug <target>\" in usage and \"$0 hugs $1.\" in response.\n\n"
            + "By default the commands are created persistent, which means they will be created again on start.\n"
            + "Non-persistent commands are lost once the bot stops or restarts.";

    private final NuclearBotGUI m_gui;

    private final JComboBox<String> m_commandCombo;
    private final JTextField m_usageField;
    private final JTextField m_descriptionField;
    private final JTextField m_responseField;

    private final UserCommandManager m_commands;

    public CommandEditPanel(final NuclearBotGUI gui)
    {
        super(new VerticalLayout());
        setBorder(BorderFactory.createTitledBorder("Create/Update"));

        m_gui = gui;

        final JPanel namePanel = new JPanel(new FlowLayout());
        {
            final JLabel namePrefixLabel = new JLabel("<html><u>Name:</u></html>");
            m_commandCombo = new JComboBox<>();

            m_commandCombo.setEditable(true);

            m_commandCombo.addActionListener(e -> updateCommandInfo());
            m_commandCombo.addItemListener(e ->
            {
                if (e.getStateChange() == ItemEvent.SELECTED)
                    updateCommandInfo();
            });
            m_commandCombo.addFocusListener(new FocusAdapter() {
                @Override
                public void focusLost(FocusEvent e)
                {
                    updateCommandInfo();
                }
            });

            namePanel.add(namePrefixLabel);
            namePanel.add(m_commandCombo);
        }

        final JPanel usagePanel = new JPanel(new FlowLayout());
        {
            final JLabel usagePrefixLabel = new JLabel("<html><u>Usage:</u></html>");
            m_usageField = new JTextField(12);

            usagePanel.add(usagePrefixLabel);
            usagePanel.add(m_usageField);
        }

        final JPanel descriptionPanel = new JPanel(new FlowLayout());
        {
            final JLabel descriptionPrefixLabel = new JLabel("<html><u>Description:</u></html>");
            m_descriptionField = new JTextField(16);

            descriptionPanel.add(descriptionPrefixLabel);
            descriptionPanel.add(m_descriptionField);
        }

        final JPanel responsePanel = new JPanel(new FlowLayout());
        {
            final JLabel responsePrefixLabel = new JLabel("<html><u>Response:</u></html>");
            m_responseField = new JTextField(16);

            responsePanel.add(responsePrefixLabel);
            responsePanel.add(m_responseField);
        }

        final JPanel submitPanel = new JPanel(new FlowLayout());
        {
            final JButton createButton = new JButton("Create/Update");
            final JButton removeButton = new JButton("Remove");
            final JButton helpButton = new JButton("Help...");

            createButton.addActionListener(e -> createCommand());

            removeButton.addActionListener(e -> removeCommand());

            helpButton.addActionListener(e ->
                    m_gui.getDialogs().info(HELP_TEXT, "Command creation help"));

            submitPanel.add(createButton);
            submitPanel.add(removeButton);
            submitPanel.add(helpButton);
        }

        m_commands = new UserCommandManager(gui, m_commandCombo);

        add(namePanel);
        add(usagePanel);
        add(descriptionPanel);
        add(responsePanel);
        add(submitPanel);

        m_commands.loadCommands();
    }

    public void registerCommands()
    {
        m_commands.registerCommands();
    }

    private void updateCommandInfo()
    {
        final String label = String.valueOf(m_commandCombo.getSelectedItem()).trim().toLowerCase();
        final UserCommandManager.CommandInfo command = m_commands.getCommand(label);
        m_usageField.setText(command != null ? command.usage : "");
        m_descriptionField.setText(command != null ? command.description : "");
        m_responseField.setText(command != null ? command.response : "");
    }

    private void createCommand()
    {
        final String name = ((String) m_commandCombo.getSelectedItem()).trim().toLowerCase();
        final String usage = m_usageField.getText().trim();
        final String description = m_descriptionField.getText().trim();
        final String response = m_responseField.getText().trim();

        m_commands.createUserCommand(name, usage, description, response, false);
    }

    private void removeCommand()
    {
        m_commands.removeCommand(((String) m_commandCombo.getSelectedItem()).trim().toLowerCase());
    }

}

File: src/nuclearbot/gui/components/commands/CommandEditPanel.kt
package nuclearbot.gui.components.commands

import nuclearbot.gui.NuclearBotGUI
import nuclearbot.gui.commands.UserCommandManager
import nuclearbot.gui.commands.UserCommandModeration
import nuclearbot.gui.utils.VerticalLayout
import nuclearbot.plugin.CommandExecutor
import java.awt.FlowLayout
import java.awt.event.FocusAdapter
import java.awt.event.FocusEvent
import java.awt.event.ItemEvent
import javax.swing.BorderFactory
import javax.swing.JButton
import javax.swing.JComboBox
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField

/*
 * Copyright (C) 2017 NuclearCoder
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * The GUI panel for the command edition panel.<br></br>
 * <br></br>
 * NuclearBot (https://github.com/NuclearCoder/nuclear-bot/)<br></br>

 * @author NuclearCoder (contact on the GitHub repo)
 */
class CommandEditPanel(private val gui: NuclearBotGUI) : JPanel(VerticalLayout()) {

    private val commandCombo = JComboBox<String>().apply {
        isEditable = true

        addActionListener { updateCommandInfo() }
        addItemListener { e ->
            if (e.stateChange == ItemEvent.SELECTED)
                updateCommandInfo()
        }
        addFocusListener(object : FocusAdapter() {
            override fun focusLost(e: FocusEvent) {
                updateCommandInfo()
            }
        })
    }

    private val usageField = JTextField(12)
    private val descriptionField = JTextField(16)
    private val responseField = JTextField(16)

    private val commands: UserCommandManager

    private val modCommands: CommandExecutor

    init {
        border = BorderFactory.createTitledBorder("Create/Update")

        val namePanel = JPanel(FlowLayout()).apply {
            add(JLabel("<html><u>Name:</u></html>"))
            add(commandCombo)
        }

        val usagePanel = JPanel(FlowLayout()).apply {
            add(JLabel("<html><u>Usage:</u></html>"))
            add(usageField)
        }

        val descriptionPanel = JPanel(FlowLayout()).apply {
            add(JLabel("<html><u>Description:</u></html>"))
            add(descriptionField)
        }

        val responsePanel = JPanel(FlowLayout()).apply {
            add(JLabel("<html><u>Response:</u></html>"))
            add(responseField)
        }

        val submitPanel = JPanel(FlowLayout()).apply {
            val createButton = JButton("Create/Update").apply { addActionListener { createCommand() } }
            val removeButton = JButton("Remove").apply { addActionListener { removeCommand() } }
            val helpButton = JButton("Help...").apply { addActionListener { gui.dialogs.info(HELP_TEXT, "Command creation help") } }

            add(createButton)
            add(removeButton)
            add(helpButton)
        }

        commands = UserCommandManager(gui, commandCombo)
        modCommands = UserCommandModeration(commands)

        add(namePanel)
        add(usagePanel)
        add(descriptionPanel)
        add(responsePanel)
        add(submitPanel)

        commands.loadCommands()
    }

    fun registerCommands() {
        gui.client.run {
            registerCommand("cmdadd", modCommands).apply {
                usage = "!cmdadd <name> <response>"
                description = "Adds a command with the given name and response."
            }

            registerCommand("cmdrem", modCommands).apply {
                usage = "!cmdrem <name>"
                description = "Removes a command with the given name."
            }

            registerCommand("cmdusage", modCommands).apply {
                usage = "!cmdusage <name> <usage>"
                description = "Sets the command usage."
            }

            registerCommand("cmddesc", modCommands).apply {
                usage = "!cmddesc <name> <usage>"
                description = "Sets the command description."
            }
        }

        commands.registerCommands()
    }

    private fun updateCommandInfo() {
        val label = commandCombo.selectedItem.toString().trim().toLowerCase()
        val command = commands.getCommand(label)

        usageField.text = command?.usage ?: ""
        descriptionField.text = command?.description ?: ""
        responseField.text = command?.response ?: ""
    }

    private fun createCommand() {
        val name = (commandCombo.selectedItem as String).trim().toLowerCase()
        if (name.isNotEmpty()) {
            val usage = usageField.text.trim()
            val description = descriptionField.text.trim()
            val response = responseField.text.trim()

            commands.createUserCommand(name, usage, description, response, false)
        }
    }

    private fun removeCommand() {
        val name = (commandCombo.selectedItem as String).trim().toLowerCase()
        if (name.isNotEmpty()) {
            commands.removeCommand(name, false)
        }
    }

    companion object {

        private const val HELP_TEXT =
                """The convention for usage format is the following:
- angle brackets for required arguments;
- square brackets for optional arguments;
but the GUI creator doesn't support optional arguments yet.
Please only write arguments without the command in the usage field.

To include arguments in the response, use "${"$"}n" or "{${"$"}n}", where n is the argument index.
The zero-th argument is the sender's username.
For instance, to create a hug command, put "!hug <target>" in usage and "$0 hugs $1." in response.

By default the commands are created persistent, which means they will be created again on start.
Non-persistent commands are lost once the bot stops or restarts."""

    }

}
